### PR TITLE
add metrics discoverer, index manager, intelligent gateway loadbalancing logic

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <i.can.feel.gravity@gmail.com>

### Description:

this PR includes 4 changes
1. Discoverer discoveries memoryordered pods
2. Gateway now uses intelligent loadbalancing using discoverer
3. Agent now returns Index_info
4. Beta Implementation of Index Manager

### Environment:

- Golang Version: 1.13.5
- Docker Version: 19.03.5
- Kubernetes Version: 1.16.3
- NGT Version: 1.8.4

### Types of changes:

- [ ] Bug fix [type/bug]
- [x] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensure all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/discoverer/k8s/router/router.go [hunk 1]
```diff
@@ -58,12 +58,20 @@ func New(opts ...Option) http.Handler {
 				h.Index,
 			},
 			{
-				"Discover",
+				"Pods",
 				[]string{
 					http.MethodPost,
```
Comment (rinx) [inline pkg/discoverer/k8s/router/router.go]: badge
It is not consistent with apis/proto/discoverer/discoverer.proto. it should be MethodGet.

File: pkg/discoverer/k8s/router/router.go [hunk 2]
```diff
@@ -58,12 +58,20 @@ func New(opts ...Option) http.Handler {
...
 				},
-				"/discoverer",
-				h.Discover,
+				"/pods",
+				h.Pods,
+			},
+			{
+				"Nodes",
+				[]string{
+					http.MethodPost,
```
Comment (rinx) [inline pkg/discoverer/k8s/router/router.go]: badge
same

File: pkg/discoverer/k8s/service/nodemap.go [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Copyright 2016 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
```
Comment (rinx) [inline pkg/discoverer/k8s/service/nodemap.go]: badge
BSD license :ok_hand:

File: pkg/discoverer/k8s/service/nodemetricsmap.go [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Copyright 2016 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
```
Comment (rinx) [inline pkg/discoverer/k8s/service/nodemetricsmap.go]: badge
BSD :ok_hand:

File: pkg/discoverer/k8s/service/podmetricsmap.go [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Copyright 2016 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
```
Comment (rinx) [inline pkg/discoverer/k8s/service/podmetricsmap.go]: badge
BSD :ok_hand:

File: pkg/discoverer/k8s/service/podsmap.go [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Copyright 2016 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
```
Comment (rinx) [inline pkg/discoverer/k8s/service/podsmap.go]: badge
BSD :ok_hand:

**Comment by kpango** [comment]
@rinx @kmrmt @hlts2 @vankichi please re-review this PR

File: internal/net/grpc/pool.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+func (c *ClientConnPool) Do(f func(conn *ClientConn) error) (err error) {
+	conn := c.Get()
+	err = f(conn)
+	c.Put(conn)
+	return err
+}
+
+func (c *ClientConnPool) Healthy() bool {
```
Comment (rinx) [inline internal/net/grpc/pool.go]: badge
`IsHealthy` is better naming.

Comment (kpango) [reply internal/net/grpc/pool.go]: fixed! thanks

File: internal/net/grpc/pool.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+	port    string
+	size    uint64
+	length  uint64
+	dopts   []DialOption
+	closing atomic.Value
+}
+
+func NewPool(ctx context.Context, addr string, size uint64, dopts ...DialOption) (*ClientConnPool, error) {
+	if !strings.HasPrefix(addr, "::") && strings.HasPrefix(addr, ":") {
+		addr = "127.0.0.1" + addr
```
Comment (rinx) [inline internal/net/grpc/pool.go]: badge
"127.0.0.1" appears more than three times in this file. it should be const, i think.

Comment (kpango) [reply internal/net/grpc/pool.go]: fixed!

File: vald/values.yaml [hunk 9]
```diff
@@ -269,7 +270,9 @@ gateway:
     discoverer:
       duration: "2s"
       discover_client: {}
-      agent_client: {}
+      agent_client: {
```
Comment (rinx) [inline vald/values.yaml]: badge
I think it doesn't needed to use "{}" here.

**Comment by rinx** [comment]
Others OK. looks good to me :+1:

File: vald/values.yaml (hunk 9 above)
Comment (kpango) [reply vald/values.yaml]: I think we need, we have to control agent_client's connection pool size

Comment (rinx) [reply vald/values.yaml]: my sugggestion is just like this:
```suggestion
      agent_client: 
         connection_pool: 100
```

Comment (kpango) [reply vald/values.yaml]: I think we doesn't need L273's { and L275's }

Comment (rinx) [reply vald/values.yaml]: Yes.

Comment (kpango) [reply vald/values.yaml]: yes, I fixed like that

Comment (kpango) [reply vald/values.yaml]: thankyou

**Comment by kpango** [comment]
Do I have to add helm implementation for index manager in this PR?

**Comment by kpango** [comment]
Or anyone can work on it?

**Comment by rinx** [comment]
Could you please send an another PR for it?
I don't want make this PR bigger :rofl:.

I can add Helm templates for index manager.

**Comment by kpango** [comment]
Certainly.
I agree with that, I'm so tired and now I can't understand what feature is this PR 😭

Sounds great please work on index manager's helm and CI
After this PR I have to work on replication manager

**Comment by rinx** [comment]
Okay. :+1:
/approve

**Comment by kpango** [comment]
let's merge it

**Pull request merged**